Add _unitrep attr of SphericalRepresentation

Cadair opened the pull request:
_unitrep allows subclasses of SphericalRepresntation to define their
UnitSphericalRepresentation equivalent.

The copy argument to relalize_frame allows in place changes to the data
attached to the frame. This provides a way to create a frame and change
the data at high speed.

eteq commented:
This may require some discussion because it includes a subtle but significant change: adding the ``copy=False`` option means we're providing a way to break the guarantee of immutability for coordinate objects.  @taldcroft @adrn @astrofrog

The ``_unitrep`` stuff is essentially independent of that, right @Cadair?

The background for this is that @Cadair was working with @ejeschke to try to speed up Ginga, and there's a need to quickly (in a GUI context) "re-populate" coordinate objects with new data, avoiding the object creation overhead.

Also, btw, @Cadair, this needs a changelog entry.

adrn commented:
Yea, the `copy` kwarg could require some thought. I don't have a strong opinion about maintaining immutability or not, but it would be a significant change. I worry slightly that it could be tricky to protect against things that would break if we open up that door...

taldcroft commented:
Was there a specific need to swap the order of UnitSphericalRepresentatation and SphericalRepresentatation?  This just makes some irrelevant diffs, or else obscures real diffs.

Cadair commented:
If I moved the _unitrep stuff out of this, would people be willing to put it into a bug fix release seeing how it's not API breaking? (Just asking because it really breaks solar coords in Ginga).

On the topic of copy kwarg. It is by far the easiest way to speed up the realize frame call (which really needs to happen for Ginga) the other option is going mad on the `BaseFrame__init__`.

I personally think that it should be ok, people (like me) will hack around it anyway.

Cadair commented:
@taldcroft I swapped them because otherwise the unitrep was not defined when creating the class attr.

taldcroft commented on astropy/coordinates/baseframe.py:
```diff
@@ -756,16 +762,25 @@ def realize_frame(self, representation):
+            return self.__class__(representation, **frattrs)
+
+        else:
+            self._set_data(representation)
```
Do you need `self._rep_cache = dict()` first to clear the cache from any other reprs that might be there?

taldcroft commented:
I think it would be better to leave `realize_frame` alone and have a new method like `update_representation_data`.  That way it is more obvious that an in-place update is being done.  We could also have a really big user-beware warning in that method that says there might be subtle issues from doing this and so it should be used with caution.

In general I think we need to open this box to support use cases like this where the current performance is not acceptable.

eteq commented:
:+1: to @taldcroft's suggestion here of a different name with a prominent ``.. warning::`` section in its docs.

@Cadair - I also like your suggestion of splitting out the `_unitrep` stuff - at the very least that makes it easier to review which changes are for what.

Cadair commented:
This is now just the `._unitrep` stuff.

Cadair commented on astropy/coordinates/representation.py:
```diff
@@ -398,53 +363,78 @@ def from_cartesian(cls, cart):
     recommended_units = {'lon': u.deg, 'lat': u.deg}
 
-    def __init__(self, lon, lat, copy=True):
+    _unitrep = UnitSphericalRepresentation
```
This line is the only actual change in here.

embray commented on CHANGES.rst:
```diff
@@ -23,6 +23,10 @@ New Features
 
+  - `SphericalRepresentation` now has a `._unitrep` class attribute to specify 
+    an equivalent UnitSphericalRepresentation. This allows subclasses of 
+    representations to pair up correctly.
```
Should this really go here if it's describing an attribute for internal-only use, or is this something authors of `SphericalRepresentation` subclasses need to be aware of?

Cadair replied:
I did not know where best to put it in the change log, the way I have written this PR does assume that all `SphericalRepresentation` classes will have a `_unitrep` attribute now, so any authors will need to update.

Should I add a catch in the `BaseCoordinateFrame.__init__` for non-updated subclasses, or are the SunPy ones the only subclasses in existence?!

embray replied:
I would move this to the "API changes" section then, if nothing else, since it's more of an internal API feature.

Also some nitpicks--use double-backticks insgead of singe-backticks, and add a reference to the PR # :)

taldcroft commented:
@Cadair - are there any diffs in `SphericalRepresentation` or `UnitSphericalRepresentation` apart from the `_unitrep` class attribute?  Difficult to tell because of the order swap of class definitions.

Also kindof a nit, but the coordinates package tends to use more verbose names, so `_unit_representation` might be more consistent.  Since it only gets used in one place the verbosity shouldn't be a problem.

Cadair commented:
@taldcroft I have changed `_unitrep` to `_unit_representation`.

The only change in `representation.py` is the addition of the attribute to the `SphericalRepresentation` class, and the order swap.

@embray I have moved the changelog entry, I think it is ok now.

embray commented:
:+1:

Cadair commented:
Any more comments on this?

taldcroft commented:
@Cadair - if I do the following there are a lot of matches in the coordinates code:
```
find astropy/coordinates -name '*.py' | xargs grep UnitSphericalRepresentation
```
So my question is whether some of these need to be generalized to use the new class attribute so that subclasses actually behave as expected.  One obvious case is in the `__repr__` method of `baseframe.py`.  This does:
```
                if (issubclass(self.representation, SphericalRepresentation) and
                        isinstance(self.data, UnitSphericalRepresentation)):
                    data = self.represent_as(self.data.__class__,
                                             in_frame_units=True)
                else:
                    data = self.represent_as(self.representation,
                                             in_frame_units=True)
```
If it is the intent (or requirement) that custom `_unit_representation` classes be a subclass of `UnitSphericalRepresentation` then this is OK, otherwise not.  Apart from that I haven't gone through and looked at each instance of `UnitSphericalRepresentation` in the code, but maybe you need to, or at least have justification for why no generalizations are needed.

This is leading into adding some tests of having a custom `_unit_representation` and putting that object through the paces.

Cadair commented:
@taldcroft I was working on the assumption that a class in the `_unit_representation` attribute would be required to be a subclass of `UnitSphericalRepresentation` are you ok with this? I will document it somewhere.

I will try and find some time to work more on this, but I am rather busy.

taldcroft commented:
OK, it the `_unit_representation` needs to be a subclass then part of my question is answered.  But I still suspect there are many cases where `UnitSphericalRepresentation` has been hardwired and assumed to be the *only* possible `UnitSphericalRepresentation` class.  E.g.:
```
if self.data.__class__ == UnitSphericalRepresentation:
...
```
And in `representation.py`:
```
...
elif other_class is UnitSphericalRepresentation:
...
return UnitSphericalRepresentation(lon=self.lon, lat=self.lat)
```
No worries on time.  This is your feature request after all!

embray commented:
Is this ready to merge, then?

taldcroft commented:
I have an outstanding question about which (if any) other instances of `UnitSphericalRepresentation(..)` in the code need to be updated to use this class configurable option.

Cadair commented:
@taldcroft I went through the code and I think I got most of the cases where it was assumed there would never be a subclass of `UnitSphericalRepresentation` or `SphericalRepresentation`.

taldcroft commented:
@Cadair - looks good but I don't see any tests of the new functionality.  I guess you can create minimal subclasses of `UnitSphericalRepresentation` and (e.g.) `ICRS` where the latter uses your new subclass for `_unit_representation`.  Then transform to and from this frame, and change representations within the frame, and make sure everything works like expected (in particular that the representation `data` have the expected class).

Cadair commented:
@taldcroft yeah cool, it should have some tests before it is merged.

astrofrog commented:
@Cadair - any chance you can add some tests today, or shall we delay this to 1.2?

Cadair commented:
@astrofrog go on then. Let me just release SunPy 0.6.1.

Cadair commented:
@astrofrog @taldcroft I added two tests that I think check everything. Although I have probably missed something.

astrofrog commented:
@Cadair - looks like there are some failures, could you look into it today?

Cadair commented:
ok so that makes no sense... I will inestigate

Cadair commented:
Ok, after some investiagtion I can not work this out.

Why is the `KeyError` being generated with the key as the whole class? I thought the keys to that dict were the string name of the Representation classes?!

My understanding of the inside of coordinates has clearly atrophied.

astrofrog commented:
@eteq - any ideas?

astrofrog commented:
@Cadair - I'm looking into this. This is due to a side-effect from another test I think.

astrofrog commented:
Basically this is due to a conflict of representation classes with the test ``test_subclass_representation``. We should make sure tests return things to the original state though, so looking into that.

astrofrog commented:
@Cadair - fix in https://github.com/Cadair/astropy/pull/17 - can you merge it in?

astrofrog commented on astropy/coordinates/representation.py:
```diff
@@ -42,8 +42,12 @@ def __init__(cls, name, bases, dct):
+        if repr_name in REPRESENTATION_CLASSES:
+            raise ValueError("Representation class {0} already defined".format(repr_name))
+
+        REPRESENTATION_CLASSES[repr_name] = cls
```
@eteq - any objections to this? This will not allow two representations named the same to be defined - that's ok, right? Or would a warning be more appropriate?

eteq replied:
I think that's just fine.  It's sort of implicitly enforced by the existence of `REPRESENTATION_CLASSES`, anyway.  Probably better to *explicitly* require this.  

The one downside is that it prevents, say, user-level code from replacing `SphericalCoordinates` with a custom class.  But that's probably *good*, as it might have unintended consequences.  Better to force them to call it `FancySphericalCoordinates`/`'fancyspherical'` or something

astrofrog replied:
Yes, I was thinking that. I don't see any good reason to override existing ones as it's hard to predict side-effects.

astrofrog commented:
@Cadair - just a couple of Sphinx warnings left. Can you fix them?

mhvk commented on CHANGES.rst:
```diff
@@ -38,6 +38,9 @@ New Features
 
   - Added ``Supergalactic`` frame to support de Vaucouleurs supergalactic
     coordinates. [#3892]
+  - ``SphericalRepresentation`` now has a ``._unitrep`` class attribute to specify 
```
This should be `_unit_representation`

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -398,53 +367,76 @@ def from_cartesian(cls, cart):
+    def represent_as(self, other_class):
+        # Take a short cut if the other clsss is a spherical representation
+        if issubclass(other_class, PhysicsSphericalRepresentation):
+            return other_class(phi=self.lon,theta=90 * u.deg - self.lat, r=1.0)
```
PEP8, should have space after `self.lon,`

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -354,23 +345,7 @@ def lat(self):
-            return UnitSphericalRepresentation(lon=self.lon, lat=self.lat)
-        else:
-            return super(SphericalRepresentation, self).represent_as(other_class)
+    # TODO: implement represent_as for efficient transformations
```
`represent_as` is in fact defined a bit further down, so remove this comment? (Or clarify it and move it further down).

mhvk commented:
See minor cleanup comments; they can be a `[skip ci]` commit to speed this up.

eteq commented on CHANGES.rst:
```diff
@@ -38,6 +38,9 @@ New Features
     coordinates. [#3892]
+  - ``SphericalRepresentation`` now has a ``._unitrep`` class attribute to specify 
+    an equivalent UnitSphericalRepresentation. This allows subclasses of 
+    representations to pair up correctly.
```
Add `` [#3757]`` at the end here.

eteq commented on CHANGES.rst:
```diff
@@ -342,6 +345,10 @@ API changes
 
 - ``astropy.coordinates``
 
+  - ``SphericalRepresentation`` now has a ``._unit_representation`` class attribute to specify 
```
I don't think you need this section here.  It's not obvious that this is what it means, but ``API Changes`` here are generally changes to *existing* pieces of the API.  This is instead an *addition*, so it only need go in "new features".

eteq commented:
Two minor comments above, and one other thing best described via comment: Can you add just a bit in the ``representations.rst`` file mentioning this attribute's existence?

Cadair commented:
I will get to this asap.

Cadair commented:
@mhvk @astrofrog @eteq I *think* this is ready!

eteq commented:
Alright, I think we can merge this once/if the tests pass!

astrofrog commented:
:+1:

Cadair commented:
Thanks everyone.
